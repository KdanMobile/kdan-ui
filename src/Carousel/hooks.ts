import { useReducer } from 'react';

enum ActionTypes {
  PREV = 'PREV',
  NEXT = 'NEXT',
  SPECIFIC = 'SPECIFIC',
  DONE = 'DONE',
}

type PrevAction = {
  type: ActionTypes.PREV;
  payload: {
    length: number;
  };
};

type NextAction = {
  type: ActionTypes.NEXT;
  payload: {
    length: number;
  };
};

type DONEAction = {
  type: ActionTypes.DONE;
};

type SpecificAction = {
  type: ActionTypes.SPECIFIC;
  payload: {
    desired: number;
  };
};

type Action = PrevAction | NextAction | SpecificAction | DONEAction;

type State = {
  desired: number;
  active: number;
  transitionEnabled: boolean;
};

const combineReducers = (
  baseReducer: (state: State, action: Action) => State,
  ...restReducers: ((state: State, action: Action) => State | undefined)[]
) => {
  return (state: State, action: Action) => {
    for (const reducer of restReducers) {
      const result = reducer(state, action);
      if (result !== undefined) return result;
    }
    return baseReducer(state, action);
  };
};

const loopReducer = (state: State, action: Action) => {
  if (action.type === ActionTypes.PREV && state.active - 1 < 0) {
    return {
      ...state,
      desired: action.payload.length - 1,
      transitionEnabled: true,
    };
  }
  if (action.type === ActionTypes.NEXT && state.active + 1 > action.payload.length - 1) {
    return {
      ...state,
      desired: 0,
      transitionEnabled: true,
    };
  }
};

const carouselReducer = (state: State, action: Action) => {
  if (action.type === ActionTypes.PREV) {
    return {
      ...state,
      active: state.active - 1,
      desired: state.active - 1,
      transitionEnabled: true,
    };
  }
  if (action.type === ActionTypes.NEXT) {
    return {
      ...state,
      active: state.active + 1,
      desired: state.active + 1,
      transitionEnabled: true,
    };
  }
  if (action.type === ActionTypes.SPECIFIC)
    return {
      ...state,
      active: action.payload.desired,
      desired: action.payload.desired,
      transitionEnabled: true,
    };
  if (action.type === ActionTypes.DONE) {
    return {
      ...state,
      active: state.desired,
      transitionEnabled: false,
    };
  }
  throw new Error('Unhandled type in carouselReducer');
};

const calcTranslate = ({
  state,
  transitionTime = 400,
  displayCount,
}: {
  state: State;
  transitionTime: number;
  displayCount: number;
}) => {
  const smooth = `transform ${transitionTime}ms ease`;
  // The value determine the direction of translate
  const direction = Math.sign(state.desired - state.active);
  // The translate distance
  const shift = (100 / displayCount) * direction;

  return {
    transition: smooth,
    transform: `translateX(${shift}%)`,
  };
};

const calcStyle = ({
  state,
  transitionTime,
  displayCount,
}: {
  state: State;
  transitionTime: number;
  displayCount: number;
}) => {
  return {
    transform: 'translateX(0)',
    // The actual moving style of item.
    left: `-${(state.active + 1) * (100 / displayCount)}%`,
    ...(state.transitionEnabled && state.active === state.desired && { transition: `left ${transitionTime}ms` }),
    ...(state.transitionEnabled &&
      state.active !== state.desired &&
      calcTranslate({ state, transitionTime, displayCount })),
  };
};

interface CarouselProps {
  loop: boolean;
  length: number;
  displayCount: number;
}

const useCarousel = ({ loop, length, displayCount }: CarouselProps) => {
  const reducer = loop ? combineReducers(carouselReducer, loopReducer) : carouselReducer;
  const [state, dispatch] = useReducer(reducer, {
    desired: 0,
    active: 0,
    transitionEnabled: false,
  });
  // The derived state define the transition style of carousel.
  const style = calcStyle({ state, displayCount, transitionTime: 400 });
  const getCarouselProps = () => ({ style, onTransitionEnd: () => dispatch({ type: ActionTypes.DONE }) });
  const onPrevClick = (length: number) => () => dispatch({ type: ActionTypes.PREV, payload: { length } });
  const onNextClick = (length: number) => () => dispatch({ type: ActionTypes.NEXT, payload: { length } });
  const onSpecificClick = (index: number) => () =>
    dispatch({ type: ActionTypes.SPECIFIC, payload: { desired: index } });
  const getPrevBtnProps = () => ({ onClick: onPrevClick(length), prev: true });
  const getNextBtnProps = () => ({ onClick: onNextClick(length), next: true });
  const getSpecificBtnProps = ({ index }: { index: number }) => ({
    active: index === state.active ? true : false,
    onClick: onSpecificClick(index),
  });
  return {
    getCarouselProps,
    getPrevBtnProps,
    getNextBtnProps,
    getSpecificBtnProps,
  };
};

export { useCarousel, loopReducer, carouselReducer };
