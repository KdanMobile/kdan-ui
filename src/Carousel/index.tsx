import React, { useEffect, useState, useRef } from "react";

import { LeftArrow, RightArrow, LeftArrow768, RightArrow768 } from "../Icon";

import {
  Wrapper,
  LeftBtn,
  RightBtn,
  ArrowBtn,
  ElementsContainer,
  ElementsOuter,
  ElementWrapper,
  Dot,
  DotWrapper,
} from "./styled";

export type CarouselProps = {
  children: React.ReactNode[];
  mode?: "dark" | "light";
  showDot?: boolean;
  showBorder?: boolean;
  showShadow?: boolean;
  displayCount?: number;
  loop?: boolean;
};

const Carousel: React.FC<CarouselProps> = ({
  children,
  mode = "dark",
  showBorder = false,
  showDot = false,
  showShadow = false,
  displayCount = 1,
  loop = false,
}: CarouselProps) => {
  const [currentIndex, setCurrentIndex] = useState(0);
  const [moveDistance, setMoveDistance] = useState(0);
  const [slides, setSlides] = useState<React.ReactNode[]>([]);
  const [transitionEnabled, setTransitionEnabled] = useState(false);
  const elementRef = useRef<HTMLDivElement>(null);
  const tempIndex = useRef(0);

  const getEleWidth = () => {
    if (elementRef.current) {
      return elementRef.current.clientWidth / displayCount;
    }
    return 0;
  };

  const slideGenerator = (start: number, end: number) => {
    const arr = [];
    for (let i = start; i <= end; i++) {
      if (i >= children.length) {
        arr.push(children[i - children.length]);
      } else {
        arr.push(children[i]);
      }
    }
    setSlides(arr);
  };

  const handleMove = (step: number) => {
    const eleWidth = getEleWidth();
    setMoveDistance((current) => current + -1 * step * eleWidth);
  };

  const handleNextDown = () => {
    if (slides.length > displayCount) return;

    setCurrentIndex((current) => {
      slideGenerator(current, current + displayCount);
      if (loop && current === children.length - 1) {
        tempIndex.current = 0;
        return 0;
      }
      if (current < children.length - 1) {
        tempIndex.current = current + 1;
        return current + 1;
      }
      tempIndex.current = current;
      return current;
    });
  };

  const handleNextUp = () => {
    if (transitionEnabled) return;

    setTransitionEnabled(true);
    handleMove(1);

    setTimeout(() => {
      slideGenerator(tempIndex.current, tempIndex.current + displayCount - 1);
      setMoveDistance(0);
      setTransitionEnabled(false);
    }, 310);
  };

  const handlePrevDown = () => {
    if (slides.length > displayCount) return;

    handleMove(1);

    setCurrentIndex((current) => {
      if (loop && current === 0) {
        const targetIndex = children.length - 1;
        slideGenerator(targetIndex, targetIndex + displayCount);
        tempIndex.current = targetIndex;
        return targetIndex;
      }
      if (current > 0) {
        const targetIndex = current - 1;
        tempIndex.current = targetIndex;
        slideGenerator(targetIndex, targetIndex + displayCount);
        return targetIndex;
      }
      tempIndex.current = 0;
      return 0;
    });
  };

  const handlePrevUp = () => {
    if (transitionEnabled) return;

    setTransitionEnabled(true);
    handleMove(-1);

    setTimeout(() => {
      slideGenerator(tempIndex.current, tempIndex.current + displayCount - 1);
      setTransitionEnabled(false);
    }, 310);
  };

  const handleDotDown = (targetIndex: number) => {
    if (slides.length > displayCount) return;

    tempIndex.current = currentIndex;

    if (currentIndex > targetIndex) {
      slideGenerator(targetIndex, currentIndex + displayCount - 1);
      handleMove(currentIndex - targetIndex);
    }
    if (currentIndex < targetIndex) {
      slideGenerator(currentIndex, targetIndex + displayCount - 1);
      setMoveDistance(0);
    }

    setCurrentIndex(targetIndex);
  };

  const handleDotUp = (targetIndex: number) => {
    if (transitionEnabled) return;

    setTransitionEnabled(true);

    if (targetIndex > tempIndex.current) {
      handleMove(targetIndex - tempIndex.current);
    } else {
      setMoveDistance(0);
    }

    setTimeout(() => {
      slideGenerator(targetIndex, targetIndex + displayCount - 1);
      setMoveDistance(0);
      setTransitionEnabled(false);
    }, 310);
  };

  useEffect(() => {
    slideGenerator(0, 0 + displayCount - 1);
  }, []);

  return (
    <Wrapper>
      {children.length > 1 && (
        <LeftBtn
          data-testid="prev-btn"
          mode={mode}
          isDisabled={loop ? false : currentIndex === 0}
          onMouseDown={handlePrevDown}
          onMouseUp={handlePrevUp}>
          {LeftArrow && <LeftArrow />}
        </LeftBtn>
      )}
      <ElementsContainer
        ref={elementRef}
        showBorder={showBorder}
        showShadow={showShadow}>
        <ElementsOuter
          playing={transitionEnabled}
          width={`${(100 * slides.length) / displayCount}%`}
          move={moveDistance}>
          {Array.isArray(slides) &&
            slides.map((item: React.ReactNode, index: number) => (
              <ElementWrapper key={`element_${index}`}>{item}</ElementWrapper>
            ))}
        </ElementsOuter>
      </ElementsContainer>
      {children.length > 1 && (
        <RightBtn
          data-testid="next-btn"
          mode={mode}
          isDisabled={loop ? false : currentIndex === children.length - 1}
          onMouseDown={handleNextDown}
          onMouseUp={handleNextUp}>
          {RightArrow && <RightArrow />}
        </RightBtn>
      )}
      {children.length > 1 && (
        <DotWrapper data-testid="dots">
          <ArrowBtn onMouseDown={handlePrevDown} onMouseUp={handlePrevUp}>
            {LeftArrow768 && <LeftArrow768 />}
          </ArrowBtn>
          {showDot &&
            Array.isArray(children) &&
            children.map((_, index) => (
              <Dot
                isActive={currentIndex === index}
                key={`dot_${index}`}
                onMouseDown={() => handleDotDown(index)}
                onMouseUp={() => handleDotUp(index)}
              />
            ))}
          <ArrowBtn onMouseDown={handleNextDown} onMouseUp={handleNextUp}>
            {RightArrow768 && <RightArrow768 />}
          </ArrowBtn>
        </DotWrapper>
      )}
    </Wrapper>
  );
};

export default Carousel;
