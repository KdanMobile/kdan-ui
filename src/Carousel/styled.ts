import styled from "styled-components";

import { hexToRGBA } from "../helpers/utility";
import colors from "../themes/colors";
import breakpoints from "../themes/breakpoints";
import zIndex from "../themes/zIndex";

export const Wrapper = styled.div`
  margin: auto;
  position: relative;
  width: 100%;
  min-width: 275px;
  max-width: 1080px;
  padding: 36px;
  display: inline-block;

  @media ${breakpoints.down("lg")} {
    padding: 0 0 36px;
  }
  @media ${breakpoints.down("md")} {
    padding: 0 0 32px;
  }
`;

export const ArrowBtn = styled.span`
  padding: 0 6px;
  display: none;
  cursor: pointer;
  margin: 0 14px;

  @media ${breakpoints.down("lg")} {
    display: inline-flex;
  }
`;

export const IconButton = styled.button`
  outline: none;
  border: 0;
  width: 72px;
  height: 72px;
  border-radius: 36px;
  box-shadow: 0px 3px 1px -2px rgba(0, 0, 0, 0.13),
    0px 2px 2px 0px rgba(0, 0, 0, 0.1), 0px 1px 5px 0px rgba(0, 0, 0, 0.05);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  position: absolute;
  top: calc(50% - 36px);
  z-index: ${zIndex[2]};
  display: flex;
  justify-content: center;
  align-items: center;

  ${({ isDisabled, mode }: { isDisabled?: boolean; mode: string }) =>
    isDisabled
      ? `
      background-color: ${colors.N15};
      cursor: default;
    `
      : `
      background-color: ${mode === "dark" ? colors.N100 : colors.P50};
      cursor: pointer;

      :hover {
        background-color: ${
          mode === "dark"
            ? hexToRGBA(colors.N100, 0.8)
            : hexToRGBA(colors.P50, 0.8)
        };
      }
    `}

  @media ${breakpoints.down("lg")} {
    display: none;
  }
`;

export const LeftBtn = styled(IconButton)`
  left: 0;
`;

export const RightBtn = styled(IconButton)`
  right: 0;
`;

export const ElementsContainer = styled.div<{
  showBorder: boolean;
  showShadow: boolean;
}>`
  overflow: hidden;
  width: 100%;
  height: 100%;
  margin: auto;
  border: ${({ showBorder }) =>
    showBorder ? `1px solid ${colors.N15}` : "none"};
  border-radius: 4px;
  box-shadow: ${({ showShadow }) =>
    showShadow ? "0 0 20px 0 rgba(130, 130, 130, 0.25)" : ""};
`;

export const ElementsOuter = styled.div<{
  width: string;
  move: number;
  playing: boolean;
}>`
  display: flex;
  transform: ${({ move }) => `translateX(${move}px)`};
  transition: ${({ playing }) =>
    playing ? "transform 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms" : "unset"};
  width: ${({ width }) => width};
`;

export const ElementWrapper = styled.div`
  width: 100%;
`;

export const DotWrapper = styled.div`
  display: inline-flex;
  position: absolute;
  left: 0;
  right: 0;
  bottom: -20px;
  justify-content: center;
  align-items: center;
`;

export const Dot = styled.span`
  width: 14px;
  height: 14px;
  border-radius: 7px;
  display: block;
  cursor: pointer;
  margin: 6px;

  background-color: ${({ isActive }: { isActive?: boolean }) =>
    isActive ? colors.N100 : colors.N15};
`;
