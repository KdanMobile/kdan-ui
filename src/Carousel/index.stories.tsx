import React from "react";

import Carousel, { CarouselProps } from ".";
import { Container } from "../component/Decorator";

export default {
  title: "Carousel",
  component: Carousel,
  decorators: [
    (story: () => React.ReactNode): React.ReactNode => (
      <Container padding>{story()}</Container>
    ),
  ],
  parameters: {
    docs: {
      description: {
        component: "幻燈片",
      },
    },
  },
};

const data = [
  { key: 1, content: "1" },
  { key: 2, content: "2" },
  { key: 3, content: "3" },
];

export const Basic = (props: CarouselProps): React.ReactNode => (
  <Carousel {...props}>
    {data.map((item) => (
      <p
        key={item.key}
        style={{
          minWidth: "275px",
          width: "100%",
          height: "250px",
          backgroundColor: "gray",
          fontSize: "72px",
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
          margin: 0,
        }}>
        {item.content}
      </p>
    ))}
  </Carousel>
);

Basic.args = {
  mode: "light",
  showDot: true,
  showBorder: true,
  displayCount: 2,
  loop: true,
} as CarouselProps;

export const ShowTwoItems = (props: CarouselProps): React.ReactNode => (
  <Carousel {...props}>
    {data.map((item) => (
      <p
        key={item.key}
        style={{
          minWidth: "275px",
          width: "100%",
          height: "250px",
          textAlign: "center",
          backgroundColor: "gray",
          margin: 0,
        }}>
        {item.content}
      </p>
    ))}
  </Carousel>
);

ShowTwoItems.args = {
  mode: "light",
  showDot: true,
  displayCount: 2,
} as CarouselProps;

export const OnlyOne = (props: CarouselProps): React.ReactNode => (
  <Carousel {...props}>
    {[
      <p
        key="one"
        style={{
          minWidth: "275px",
          width: "100%",
          height: "250px",
          textAlign: "center",
          backgroundColor: "gray",
          margin: 0,
        }}>
        only one
      </p>,
    ]}
  </Carousel>
);

OnlyOne.args = {
  mode: "light",
  showDot: true,
} as CarouselProps;
