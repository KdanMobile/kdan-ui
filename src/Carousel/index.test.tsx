import React from "react";
import { render, cleanup } from "@testing-library/react";
import "jest-styled-components";
import "@testing-library/jest-dom";

import Carousel from "./";

describe("Carousel", () => {
  afterEach(cleanup);

  test("match snapshot", () => {
    const { container } = render(<Carousel>{[<div key="1">1</div>]}</Carousel>);

    expect(container).toMatchSnapshot();
  });

  test("show dots", () => {
    const { getByTestId } = render(
      <Carousel showDot>
        {[<div key="1">1</div>, <div key="2">2</div>]}
      </Carousel>
    );

    expect(getByTestId("dots")).toBeInTheDocument();
  });
});
