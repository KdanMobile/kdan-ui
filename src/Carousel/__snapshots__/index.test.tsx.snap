// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel match snapshot 1`] = `
.c0 {
  margin: auto;
  position: relative;
  width: 100%;
  min-width: 275px;
  max-width: 1080px;
  padding: 36px;
  display: inline-block;
}

.c1 {
  overflow: hidden;
  width: 100%;
  height: 100%;
  margin: auto;
  border: none;
  border-radius: 4px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
  -webkit-transition: unset;
  transition: unset;
  width: 100%;
}

.c3 {
  width: 100%;
}

@media (max-width:1023.5px) {
  .c0 {
    padding: 0 0 36px;
  }
}

@media (max-width:767.5px) {
  .c0 {
    padding: 0 0 32px;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        width="100%"
      >
        <div
          class="c3"
        >
          <div>
            1
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
